nilearn/plotting/displays.py

```diff
@@ -1203,6 +1203,242 @@ def draw_cross(self, cut_coords=None, **kwargs):
+            self.frame_axes.imshow(np.zeros((2, 2, 3)),
+                                   extent=[-5000, 5000, -5000, 5000],
+                                   zorder=-500, aspect='auto')
+            self.frame_axes.set_zorder(-1000)
+
+    def _locator(self, axes, renderer):
+        """ The locator function used by matplotlib to position axes.
+            Here we put the logic used to adjust the size of the axes.
+        """
+
+        x0, y0, x1, y1 = self.rect
+        width_dict = dict()
+        height_dict = dict()
+        # A dummy axes, for the situation in which we are not plotting
+        # all three (x, y, z) cuts
+        dummy_ax = self._axes_class(None, None, None)
+        width_dict[dummy_ax.ax] = 0
+        height_dict[dummy_ax.ax] = 0
+        display_ax_dict = self.axes
+
+        if self._colorbar:
+            adjusted_width = self._colorbar_width / len(self.axes)
+            right_margin = self._colorbar_margin['right'] / len(self.axes)
+            ticks_margin = self._colorbar_margin['left'] / len(self.axes)
+            x1 = x1 - (adjusted_width + ticks_margin + right_margin)
+
+        for display_ax in display_ax_dict.values():
+            bounds = display_ax.get_object_bounds()
+            if not bounds:
+                # This happens if the call to _map_show was not
+                # succesful. As it happens asyncroniously (during a
+                # refresh of the figure) we capture the problem and
+                # ignore it: it only adds a non informative traceback
+                bounds = [0, 1, 0, 1]
+            xmin, xmax, ymin, ymax = bounds
+            width_dict[display_ax.ax] = (xmax - xmin)
+            height_dict[display_ax.ax] = (ymax - ymin)
+
+        #
+        unique_width_dict = {}
+
+        for key,value in width_dict.items():
+            if value not in unique_width_dict.values():
+                unique_width_dict[key] = value
+
+        unique_height_dict = {}
+
+        for key,value in height_dict.items():
+            if value not in unique_height_dict.values():
+                unique_height_dict[key] = value
+    
+        total_width = float(sum(unique_width_dict.values()))  
+
+        for ax, width in width_dict.items():
+            width_dict[ax] = width / total_width * (x1 - x0)
+
+        total_height = float(sum(unique_height_dict.values()))
+
+        for ax, height in height_dict.items():
+            height_dict[ax] = height / total_height * (y1 - y0)
+
+        direction_ax = []
+        for d in self._cut_displayed:
+            direction_ax.append(display_ax_dict.get(d, dummy_ax).ax)
+
+        coord1_dict = dict()
+        coord2_dict = dict()
+        coord3_dict = dict()
+        coord4_dict = dict()
+
+        if 'y' in self.axes:
+            ax = self.axes['y'].ax
+            coord1_dict[ax] = x0
+            coord2_dict[ax] = (y1) - height_dict[ax]
+            coord3_dict[ax] = x0 + width_dict[ax]
+            coord4_dict[ax] = y1
+
```

Comment:
I have been looking into that and I spoke with a few other engineers.
There's a higher level structure in Python StdLib designed for exactly this situation.

```
with contextlib.suppress(KeyError):
    ax = self.axes['y'].ax
    coord1[ax] = x0
    coord2[ax] = (y1) - height[ax]
    coord3[ax] = x0 + width[ax]
    coord4[ax] = y1
```
If the specified error is raised the subsequent `with` block is not executed.

The benefit of this it retains the cleanness of `if` and its not difficult to understand by new contributors - one google search away when someone looks at it the first time and they will understand & remember.
Another major benefit is that it does not add the double processing overhead of doing an `if` and then membership check using `in` on a sequence structure (`dict.keys()`).

The library is littered with such `if-in` lookups and together they add processing overhead that the standard library is well equipped to avoid.

Reply:
I agree with @jeromedockes. The snippet proposed by @kchawla-pi makes the code less readable (and violates the first 3 rules of the [zen of python](https://en.wikipedia.org/wiki/Zen_of_Python)). The rule of thumb here would use the `try/catch` if it makes sense to write 2 separate tests one just calling whatever needs to be called, and another one calling `with pytest.raises(bla)`. If that's not the case you most probably don't want to use a try catch.

what I dislike about what the author is proposing is that there's no else branch and therefore is incomplete (or degraded) conditional. Still valid python syntax but a degraded conditional case none the less (and therefore can never be cover). Thats not the case for the snipped proposed by @kchawla-pi but this can be solved by doing this:

```py
       if 'y' in self.axes:
            ax = self.axes['y'].ax
            coord1_dict[ax] = x0
            coord2_dict[ax] = (y1) - height_dict[ax]
            coord3_dict[ax] = x0 + width_dict[ax]
            coord4_dict[ax] = y1
       else:
          pass
```
But since this is rearely seen in scientific codebase and degraded conditionals are not offending most scientific users/developers. I would not even bother.



Once said all that, whatever you do for `x`, `y`, and `z` is quite similar. I'm sure you can factor it. And it seems quite a natural thing to be doing several times, so I would check other places in the code doing similar things, factor it nicely and make function out of it.

Reply:
> There's a higher level structure in Python StdLib designed for exactly this situation.
> 
> ```
> with contextlib.suppress(KeyError):
>     ax = self.axes['y'].ax
>     coord1[ax] = x0
>     coord2[ax] = (y1) - height[ax]
>     coord3[ax] = x0 + width[ax]
>     coord4[ax] = y1
> ```
> 
> If the specified error is raised the subsequent `with` block is not executed.

but then a bug could be hidden if 'y' is in axes but for example ax is not in coord1

Reply:
> Another major benefit is that it does not add the double processing overhead of doing an `if` and then membership check using `in` on a sequence structure (`dict.keys()`).

dict have hash tables so a lookup is fast. Moreover catching an exception has a  cost too. more importantly this part of the code is definitely not a performance bottleneck so readability is all that matters

Reply:
>  The snippet proposed by @kchawla-pi makes the code less readabl
Agrees, I looked at it again and didn't really find it clean, so I looked for the contextlib solution.

>  then a bug could be hidden
True, I guess we shouldn't use the contextib.suppress

> if 'y' is in axes but for example ax is not in coord1
That is not being handled anywhere, including in the way it was originally written. We should handle it and give a message that makes sense in this context.

```
      if 'y' in self.axes:
            ax = self.axes['y'].ax 
            coord1_dict[ax] = x0
            coord2_dict[ax] = (y1) - height_dict[ax]
            coord3_dict[ax] = x0 + width_dict[ax]
            coord4_dict[ax] = y1
```

Perhaps switching back to `if` is one way to do it.

>  double processing overhead of doing an if and then membership check using in on a sequence structure (dict.keys()).
I am concerned about this, Nilearn will get more complex & CircleCI build times keep going up. If we can sensibly replace the ton of `if`'s we use with more try-except, it'll be better, esp in new code. If you have ideas about that let me know.